seal.go

```diff
@@ -0,0 +1,347 @@
+	case err != nil:
+		return xerrors.Errorf("cannot create context for lock NV index: %w", err)
+	}
+
+	lockIndexPub, err := readAndValidateLockNVIndexPublic(tpm.TPMContext, lockIndex, session)
+	if err != nil {
+		return ErrTPMProvisioning
+	}
+	lockIndexName, err := lockIndexPub.Name()
+	if err != nil {
+		return xerrors.Errorf("cannot compute name of global lock NV index: %w", err)
+	}
+
+	succeeded := false
+
+	// Create destination files
+	keyFile, err := os.OpenFile(keyPath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0600)
+	if err != nil {
+		return xerrors.Errorf("cannot create key data file: %w", err)
+	}
+	defer func() {
+		keyFile.Close()
+		if succeeded {
+			return
+		}
+		os.Remove(keyPath)
+	}()
+
+	var policyUpdateFile *os.File
+	if policyUpdatePath != "" {
+		var err error
+		policyUpdateFile, err = os.OpenFile(policyUpdatePath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0600)
+		if err != nil {
+			return xerrors.Errorf("cannot create private data file: %w", err)
+		}
+		defer func() {
+			policyUpdateFile.Close()
+			if succeeded {
+				return
+			}
+			os.Remove(policyUpdatePath)
+		}()
+	}
+
+	// Create an asymmetric key for signing authorization policy updates, and authorizing dynamic authorization policy revocations.
+	authKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return xerrors.Errorf("cannot generate RSA key pair for signing dynamic authorization policies: %w", err)
+	}
+	authPublicKey := createPublicAreaForRSASigningKey(&authKey.PublicKey)
+	authKeyName, err := authPublicKey.Name()
+	if err != nil {
+		return xerrors.Errorf("cannot compute name of signing key for dynamic policy authorization: %w", err)
+	}
+
+	// Create pin NV index
+	pinIndexPub, pinIndexAuthPolicies, err := createPinNVIndex(tpm.TPMContext, params.PinHandle, authKeyName, session)
+	switch {
+	case tpm2.IsTPMError(err, tpm2.ErrorNVDefined, tpm2.CommandNVDefineSpace):
+		return TPMResourceExistsError{params.PinHandle}
+	case isAuthFailError(err, tpm2.CommandNVDefineSpace, 1):
+		return AuthFailError{tpm2.HandleOwner}
+	case err != nil:
+		return xerrors.Errorf("cannot create new pin NV index: %w", err)
+	}
+	defer func() {
+		if succeeded {
+			return
+		}
+		index, err := tpm2.CreateNVIndexResourceContextFromPublic(pinIndexPub)
+		if err != nil {
+			return
+		}
+		tpm.NVUndefineSpace(tpm.OwnerHandleContext(), index, session)
+	}()
+
+	sealedKeyNameAlg := tpm2.HashAlgorithmSHA256
+
+	// Compute the static policy - this never changes for the lifetime of this key file
+	staticPolicyData, authPolicy, err := computeStaticPolicy(sealedKeyNameAlg, &staticPolicyComputeParams{
+		key:                  authPublicKey,
+		pinIndexPub:          pinIndexPub,
+		pinIndexAuthPolicies: pinIndexAuthPolicies,
+		lockIndexName:        lockIndexName})
+	if err != nil {
+		return xerrors.Errorf("cannot compute static authorization policy: %w", err)
+	}
+
+	// Define the template for the sealed key object, using the computed policy digest
+	template := tpm2.Public{
```

Comment:
maybe this should be global,  and used to get the relevant values in the validation code?

Reply:
It doesn't really work as a global because the AuthPolicy field needs to be mutated here. I've added makeSealedKeyTemplate() though which returns a *tpm2.Public without the AuthPolicy field set.